### ISSUE TITLE
Screens being put to sleep after starting PinnballY

### ISSUE THREAD

**Issue by jsb2092** [mjrgh/PinballY#128]
Been running a b22 build I did myself for a few weeks now.  Recently, it has started putting the screens to sleep after launching PinnballY.  Namely, the Copyright screen comes up, and then it switches to display the currently table information and the playfield screen and backglass go to sleep. 

Windows power settings are set to never put the computer to sleep

After waiting up the screen, using a keyboard, PinballY will continue to function correctly.

This is repeatable with multiple launches of Pinball Y.

In this video, I launch PinnballY, the initial screen comes up, and then the screen saver kicks on.  I then hit shift of the keyboard and it comes back.

https://www.dropbox.com/s/2lshil2m7ot2bkw/IMG_2046.MOV?dl=0

**Comment by mjrgh** [comment]
Strange. I can't think of anything that PinballY is doing that would affect power settings. Did you make any changes, or maybe doing something in Javascript?

**Comment by jsb2092** [comment]
Nope, no changes.  As far as I know there were no windows updates either.  I was using the flexdmd javascript for a while, but turned that off weeks ago; and it didn't stop working when I turned it off.  It strange, because you can see in the video it happens just as it switches from the copyright (initial info screen) to the table info screen.

**Comment by mjrgh** [comment]
Could be a timing coincidence - maybe try turning off the splash screen (Startup options page) and see if it still happens.

**Comment by jsb2092** [comment]
ok, one thing did change; I got a plunger from zebsboards last week. I set it up, but it's not currently hooked up. 

I turned off DOF effects and the splash screen and it still does it.

The logs aren't terribly useful either
[PinballY.log](https://github.com/mjrgh/PinballY/files/4434633/PinballY.log)

**Comment by jsb2092** [comment]
Also, I don't actual think it's power, I think PinballY just stops outputting a signal somehow.

**Comment by jsb2092** [comment]
also, definitely related to the zebs stuff - I plugged in the plunger (and thus control board) and it doesn't do it anymore, aka, works fine.

**Comment by mjrgh** [comment]
So not a PinballY issue?

**Comment by jsb2092** [comment]
I wouldn't go that far. Something in PinballY is causing the screens to turn off when the zebs controller isn't hooked up, which isn't right. I can find nothing in the instructions (https://www.zebsboards.com/forum/ext/dmzx/fileupload/files/3671fe9144a88b2f568a6402b09006c1.pdf) that would be anything PinballY would be using directly. My guess is PinballY is trying to access something on the controller port (COM4), and not finding the device, and then going into a weird state. You know the code better than I do though, so I am not sure what happens when it launches to access the controller.

**Comment by mjrgh** [comment]
PinballY isn't trying to access anything in the controller or any COM ports.  There might be some interaction with DOF.  Given that it's isolated to the Zeb's software, though, I can't see how it's anything I can fix in PinballY.

**Comment by jsb2092** [comment]
Shouldn't turning off DOF in pinballY make it not load anything from there?

**Comment by mjrgh** [comment]
Right.

**Comment by jsb2092** [comment]
ok, so that doesn't seem to be working :(

**Comment by mjrgh** [comment]
How do you figure?

**Comment by jsb2092** [comment]
If I disabled DOF in PinballY, the issue still occurs. Since connecting the board makes the issue go away, something has to be happening, somewhere that isn't getting initialized, or run correctly when it's not hooked up. I am contacting the zebs board people, but there was no software from them that I installed, so I'm not sure how it would be affected (I did install the software the configures the board via a script, but that is run once).

So the question is, if I got rid of the zebs plunger and controller (never to use again), this issue would persist, yet there is no software to remove, uninstall, or change that came with the controller, so I'm not sure what the issue could be.

Yes, I realize I am being a pain in the ass, but i'm not sure how the issue could be out of the scope of PinballY. I could certainly try to build and debug it and see if I can find anything in the debugger.

**Comment by mjrgh** [comment]
>I am contacting the zebs board people, but there was no software from them that I installed, so I'm not sure how it would be affected (I did install the software the configures the board via a script, but that is run once).

But wait, I thought you said earlier "I set up the Zeb's board but didn't attach the device".  I took that to mean "I installed some software".  I guess it meant something else?

> So the question is, if I got rid of the zebs plunger and controller (never to use again), this issue would persist, yet there is no software to remove, uninstall, or change that came with the controller, so I'm not sure what the issue could be.

Indeed, but apparently the problem is isolated to the Zeb's device, right?  Having it plugged in makes the problem go away, having it unplugged makes the problem come back, and **never having installed it** makes the problem never show up in the first place.  As far as I can tell, Zeb's did **something** to your system - it made some kind of persistent change.  So there must be **something** to uninstall.  Again, I don't see how this is something I can fix in PinballY.  Yes, there's some interaction, but whatever that interaction is must be on the Zeb's side, it seems to me.

**Comment by jsb2092** [comment]
When I said I set it up, I plugged it into the USB port. The software is just to configure the board to put it in the correct mode.  There is no software to interface with it that I installed

The software linked here: https://www.zebsboards.com/forum/ext/dmzx/fileupload/files/058891f25a26a978a0ff37761cbd7f2c.zip is just some scripts that run to configure the device, and it registers a shell script:
CreateObject("Wscript.Shell").Run "ZBGP.bat",0,True
but there is nothing that you do that should be calling that, correct?

**Comment by mjrgh** [comment]
> but there is nothing that you do that should be calling that, correct?

Well, certainly nothing directly, since I've never heard of that before.  But "registers a shell script" presumably means that it's set up to trigger on some system event, I assume?  "Registered" means registered with Task Scheduler or something like that?  Maybe "uninstall" in this context means that you have to undo whatever that registration was.

**Comment by jsb2092** [comment]
the script doesn't actually register anything,  it simply sends data to the com port.  The plunger and control board register as a keyboard.  It does have a nudge and tilt sensor that may be messing things up, though I'm not entirely sure how.

**Comment by mjrgh** [comment]
I wish I had more ideas about what it could be.   All evidence points to there being some persistent software footprint that the Zeb's device installed somewhere along the way, and that goes a little haywire when the device isn't physically attached.  I'd think a device driver is the most likely thing if you didn't install anything explicitly. It seems to me that figuring out what that software is and how to remove it is the best bet for a solution; hopefully Zeb knows what the mystery software component is.

**Comment by jsb2092** [comment]
Since it's registered as a keyboard, I'm not sure how it could be any mystery software. In any event, the fact that it somehow causes PinballY to stop outputting video is the strange part.  Is there a default key combination in PinballY that puts it to sleep or anything like that?

**Comment by mjrgh** [comment]
> Is there a default key combination in PinballY that puts it to sleep or anything like that?

Nope; PinballY doesn't do anything with monitor power management at all.  It just displays windows.

> Since it's registered as a keyboard, I'm not sure how it could be any mystery software.

Keyboards and joysticks are (generally) handled through the built-in HID drivers, so those shouldn't be installing anything.  But you mentioned a COM port, didn't you?  That sounds like a place where a driver is generally needed.

**Comment by jsb2092** [comment]
I think I mentioned earlier, it's not going to sleep, it's just that windows just stops outputting video.  This doesn't happen if I start the tables directly.  I may have to run it through a debugger.

There are 2 parts - the COM port is to program it, but the input is registered as a keyboard.

**Comment by jsb2092** [comment]
I have been able to reproduce this with a brand new install, where I restored the tables, media and database from 6 weeks ago (before the plunger), by running in a VM, the same thing happens.  It appears to happen just after the mouse is moved off the screen (hidden).

**Comment by stanlje** [comment]
I have a KL25Z Pinscape controller and it's doing the same thing. When it's unplugged, the screens go to sleep after a couple of seconds of starting up (but not the system, you can still navigate with buttons). Plugged in and everything works like normal, no sleep. Doesn't the pinscape just emulate a joystick? I'm only using the nudge function, nothing else.

**Comment by mjrgh** [comment]
> I think I figured it out. When LauchFocus.Enabled = 1 is set, it breaks (though not always), and turning this off makes it work 100% of the time.

That's good narrowing down.  I assume "breaks" means "the screens go to sleep" (or whatever it is) and "work" means they **don't** go to sleep?

The launch focus option just synthesizes a mouse click in the main window after the launch focus delay (LaunchFocus.delay in the settings, time given in seconds).  It's hard to imagine why that would make your screens go to sleep, but if you want to look into it further, here's where you can look in the debugger to see the specific steps taken to synthesize the mouse click:

PlayfieldView.cpp -> PlayfieldView::ForceTakeFocus()

Can you think of any interaction with other software you have that might be hooking into the mouse clicks or something?  I can't think of any reason a vanilla Windows system would blank/sleep screens when clicking in the middle of a window, but maybe I'm missing something.

**Comment by jsb2092** [comment]
Yes, breaks means the screens go to sleep (though they can be woken up with a keyboard input).

As I said in a previous comment, I did a new vanilla install on a windows machine (VM) that never had a plunger or other software installed and the issue still occurred, so I'm kind of at a loss here.  I will see if I can load it up in the debugger, but I'm not sure how far I'll get as I haven't done C++ in 20 years.

**Comment by mjrgh** [comment]
> I did a new vanilla install on a windows machine (VM) that never had a plunger or other software installed

I saw that, but out of curiosity, is that VM running on the same machine where you did have the problem originally?  Maybe the problem is happening at the level outside of the VM, in the host environment.  If it's a completely virgin install on completely virgin hardware, then yes, that is pretty baffling.

**Comment by jsb2092** [comment]
The VM was on a Mac Mini, so completely different hardware, so yeah, other than the copy of the tables, nothing was the same.

**Comment by euston-r** [comment]
Hi - I have this issue too - was about to raise a bug report and saw it was already here.
I installed the latest version (beta 24) and I can confirm that the screens go to sleep on start of pinbally.
I've been able to track it down - its the "Force keyboard focus at startup" option in the startup options. Disable that, and the problem goes away. Change the delay time, and the screens go to sleep when that timer expires - so its definitely that whats causing it.

**Comment by euston-r** [comment]
I've looked at the documentation for this, and it seems its using a mouse click to focus. Perhaps the location of that click is causing the trouble...

**Comment by mjrgh** [comment]
Glad you found a solution.  The "force keyboard focus at startup" option just sends a mouse click to the main window when the timer expires, so there must be an interaction with some other program you guys have running, or maybe with one of those little monitor-edge popup controls in Windows 8 and 10.  If you guys can maybe try repeating the process a few times and see if you can see any icons popping up (it will probably be EXTREMELY quick and fleeting, so you might have to observe it many times before you even see it flicker into existence), you might be able to identify the add-on or Windows element that's getting triggered.  There must be SOMETHING external that's getting triggered by the mouse click, because mouse clicks by themselves don't make monitors go to sleep.  If you can identify what that is, maybe we can come up with a better solution than "just turn the option off".  But until then I guess the solution is "turn the option off".

**Comment by mjrgh** [comment]
I'm closing this given the workaround, and there's also a mention of it in the troubleshooting section. If anyone discovers anything new on this (like a definitive explanation of why it happens, or anything that can be done in the video card settings to prevent it), please update this so I can add the new information to the troubleshooting section.

**Comment by xantari** [comment]
I posed this question to multiple folks on the PinballY facebook group and we aren't alone, it's random when this happens.

I'm able to pretty consistently replicate this issue on my desktop computer immediately after a reboot. Then after launching it a few times the issue just goes away and I can no longer replicate it.

I've waited for all the startup tasks to finish. I've disabled the nvidia geforce experience overlay stuff (next I need to try uninstall it), and go through the process of elimination to see if I can figure it out. The pinball Y log doesn't yield any info, just mostly info about PinMemHI stuff....
